File: baboon-compiler/src/main/resources/baboon-runtime/cs/BaboonTypeMeta.cs
#nullable enable

using System;
using System.IO;
using Newtonsoft.Json.Linq;

// ReSharper disable UnusedTypeParameter
// ReSharper disable CheckNamespace
// ReSharper disable UnusedAutoPropertyAccessor.Global
// ReSharper disable MemberCanBeProtected.Global
// ReSharper disable MemberCanBePrivate.Global
// ReSharper disable ConvertToPrimaryConstructor
// ReSharper disable UnusedMember.Global
// ReSharper disable UnusedMemberInSuper.Global
// ReSharper disable ArrangeNamespaceBody
// ReSharper disable UnusedType.Global
// ReSharper disable InconsistentNaming
// ReSharper disable ClassCanBeSealed.Global

namespace Baboon.Runtime.Shared
{
    public sealed record BaboonVersion(int Major, int Minor, int Patch) : IComparable<BaboonVersion>
    {
        public int CompareTo(BaboonVersion? other)
        {
            if (other is null) return 1;
            var c = Major.CompareTo(other.Major);
            if (c != 0) return c;
            c = Minor.CompareTo(other.Minor);
            if (c != 0) return c;
            return Patch.CompareTo(other.Patch);
        }

        public static BaboonVersion From(string version)
        {
            var chunks = version.Split('.');
            if (chunks.Length < 3)
            {
                throw new Exception($"Expected to have version in format x.y.z, got {version}");
            }

            int Parse(string s, string slot)
            {
                if (!int.TryParse(s.Trim(), out var v))
                {
                    throw new Exception($"Expected to have version in format x.y.z, got {version}. Invalid {slot} value.");
                }
                return v;
            }

            return new BaboonVersion(Parse(chunks[0], "major"), Parse(chunks[1], "minor"), Parse(chunks[2], "patch"));
        }

        public static bool operator <(BaboonVersion a, BaboonVersion b) => a.CompareTo(b) < 0;
        public static bool operator >(BaboonVersion a, BaboonVersion b) => a.CompareTo(b) > 0;
        public static bool operator <=(BaboonVersion a, BaboonVersion b) => a.CompareTo(b) <= 0;
        public static bool operator >=(BaboonVersion a, BaboonVersion b) => a.CompareTo(b) >= 0;
    }

    public sealed record BaboonDomainVersion(string DomainIdentifier, string DomainVersion)
    {
        // Parses on each access; versions are small strings and the parse is cheap. Avoiding a
        // backing field keeps the positional-record shape simple and value-equality intact.
        public BaboonVersion Version => BaboonVersion.From(DomainVersion);
    }

    public sealed record BaboonTypeMeta(
        byte MetaVersion,
        string DomainIdentifier,
        string DomainVersion,
        string DomainVersionMinCompat,
        string TypeIdentifier
    )
    {
        public BaboonDomainVersion VersionRef => new BaboonDomainVersion(DomainIdentifier, DomainVersion);

        public BaboonDomainVersion? VersionMinCompat
        {
            get
            {
                if (string.IsNullOrEmpty(DomainVersionMinCompat)) return null;
                if (DomainVersionMinCompat == DomainVersion) return null;
                return new BaboonDomainVersion(DomainIdentifier, DomainVersionMinCompat);
            }
        }

        public void WriteBin(BinaryWriter writer) => BaboonTypeMetaCodec.WriteBin(this, writer);

        public JToken WriteJson() => BaboonTypeMetaCodec.WriteJson(this);

        public static BaboonTypeMeta From(IBaboonGenerated value, Type? declaredType)
        {
            // Codec discovery with ADT awareness, mirroring Scala's BaboonTypeMeta.from:
            // when the user-declared static type is the ADT trait/interface, use the ADT's
            // type identifier so the encoder can wrap with the ADT meta envelope. When the
            // user-declared type is the concrete branch, use the branch identifier directly.
            string typeIdentifier;
            if (value is IBaboonAdtMemberMeta adt && declaredType is { IsInterface: true })
            {
                typeIdentifier = adt.BaboonAdtTypeIdentifier();
            }
            else
            {
                typeIdentifier = value.BaboonTypeIdentifier();
            }

            // Codegen invariant (mirrors Scala BaboonRuntimeShared.scala:138): BaboonSameInVersions
            // is always non-empty. Index directly so a violation throws IndexOutOfRangeException
            // rather than silently masquerading as a same-version meta.
            var minCompat = value.BaboonSameInVersions()[0];

            return new BaboonTypeMeta(
                BaboonTypeMetaCodec.META_VERSION,
                value.BaboonDomainIdentifier(),
                value.BaboonDomainVersion(),
                minCompat,
                typeIdentifier
            );
        }

        public static BaboonTypeMeta? ReadMeta(BinaryReader reader) => BaboonTypeMetaCodec.ReadMeta(reader);

        public static BaboonTypeMeta? ReadMeta(JToken json) => BaboonTypeMetaCodec.ReadMeta(json);
    }

    public static class BaboonTypeMetaCodec
    {
        public const byte META_VERSION_1 = 16;
        public const byte META_VERSION = META_VERSION_1;

        public const string META_VERSION_KEY = "$mv";
        public const string DOMAIN_IDENTIFIER_KEY = "$d";
        public const string DOMAIN_VERSION_KEY = "$v";
        public const string DOMAIN_VERSION_MIN_COMPAT_KEY = "$uv";
        public const string TYPE_IDENTIFIER_KEY = "$t";

        public static void WriteBin(BaboonTypeMeta meta, BinaryWriter writer)
        {
            writer.Write(META_VERSION);
            writer.Write(meta.DomainIdentifier);
            writer.Write(meta.DomainVersion);
            if (meta.DomainVersion == meta.DomainVersionMinCompat)
            {
                writer.Write((byte)0);
            }
            else
            {
                writer.Write((byte)1);
                writer.Write(meta.DomainVersionMinCompat);
            }
            writer.Write(meta.TypeIdentifier);
        }

        public static JToken WriteJson(BaboonTypeMeta meta)
        {
            var obj = new JObject
            {
                [DOMAIN_IDENTIFIER_KEY] = meta.DomainIdentifier,
                [DOMAIN_VERSION_KEY] = meta.DomainVersion,
                [TYPE_IDENTIFIER_KEY] = meta.TypeIdentifier,
            };
            if (meta.DomainVersion != meta.DomainVersionMinCompat)
            {
                obj[DOMAIN_VERSION_MIN_COMPAT_KEY] = meta.DomainVersionMinCompat;
            }
            return obj;
        }

        public static BaboonTypeMeta? ReadMeta(BinaryReader reader)
        {
            var metaVersion = reader.ReadByte();
            if (metaVersion != META_VERSION_1) return null;

            var domainIdentifier = reader.ReadString();
            var domainVersion = reader.ReadString();
            var hasMinCompat = reader.ReadByte();
            var domainVersionMinCompat = hasMinCompat == 1 ? reader.ReadString() : domainVersion;
            var typeIdentifier = reader.ReadString();

            return new BaboonTypeMeta(
                META_VERSION,
                domainIdentifier,
                domainVersion,
                domainVersionMinCompat,
                typeIdentifier
            );
        }

        public static BaboonTypeMeta? ReadMeta(JToken json)
        {
            if (json is not JObject obj) return null;

            // Mirror Scala readMeta(json) (BaboonRuntimeShared.scala:197-205): if $mv is present
            // and not "1", reject; if absent, fall through to v1 read.
            var mvToken = obj[META_VERSION_KEY];
            if (mvToken != null)
            {
                if (mvToken.Type != JTokenType.String) return null;
                var mvStr = mvToken.Value<string>();
                if (mvStr is null) return null;
                if (!byte.TryParse(mvStr, out var mv) || mv != META_VERSION_1) return null;
            }

            var d = obj[DOMAIN_IDENTIFIER_KEY]?.Value<string>();
            var v = obj[DOMAIN_VERSION_KEY]?.Value<string>();
            var t = obj[TYPE_IDENTIFIER_KEY]?.Value<string>();

            if (d is null || v is null || t is null) return null;

            var uv = obj[DOMAIN_VERSION_MIN_COMPAT_KEY]?.Value<string>() ?? v;

            return new BaboonTypeMeta(META_VERSION, d, v, uv, t);
        }
    }
}

File: baboon-compiler/src/main/resources/baboon-runtime/rust/baboon_type_meta.rs
// BaboonTypeMeta — the top-level codec envelope (docs/spec/codec-envelope.md) and its wire codec.
// Split out of `baboon_codecs_facade.rs`, which re-exports both symbols, to keep each embedded
// runtime file under the JVM's 64KB string-constant limit.

use crate::any_opaque::BaboonCodecError;
use crate::baboon_codecs_facade::BaboonDomainVersion;

// --- BaboonTypeMeta (synthetic; used for facade lookup) ---

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BaboonTypeMeta {
    pub meta_version: u8,
    pub domain_identifier: String,
    pub domain_version: String,
    pub domain_version_min_compat: String,
    pub type_identifier: String,
    /// Oldest domain version whose JSON codec can decode the payload under the json-additive
    /// contract (tolerant key lookup; fields unknown to that version are dropped). Always
    /// <= domain_version_min_compat. Published as `$rv` when it differs from the (effective)
    /// minCompat; the binary v1 envelope does not carry it. `new` defaults it to minCompat.
    pub domain_version_readable_min: String,
}

impl BaboonTypeMeta {
    pub const META_VERSION_1: u8 = 1;
    pub const META_VERSION: u8 = Self::META_VERSION_1;
    /// Tier key of the JSON envelope's readable-min bound in `baboon_min_reader_versions_dyn`.
    pub const JSON_READABLE_TIER: &'static str = "json-additive";
    /// Tier keys of the UEBA prefix bounds in `baboon_min_reader_versions_dyn`, per index mode.
    pub const UEBA_PREFIX_COMPACT_TIER: &'static str = "prefix-compact";
    pub const UEBA_PREFIX_ANY_MODE_TIER: &'static str = "prefix-any-mode";

    pub fn new<D: Into<String>, V: Into<String>, MC: Into<String>, T: Into<String>>(
        meta_version: u8,
        domain_identifier: D,
        domain_version: V,
        domain_version_min_compat: MC,
        type_identifier: T,
    ) -> Self {
        let domain_version_min_compat: String = domain_version_min_compat.into();
        BaboonTypeMeta {
            meta_version,
            domain_identifier: domain_identifier.into(),
            domain_version: domain_version.into(),
            domain_version_readable_min: domain_version_min_compat.clone(),
            domain_version_min_compat,
            type_identifier: type_identifier.into(),
        }
    }

    pub fn with_readable_min<R: Into<String>>(mut self, readable_min: R) -> Self {
        self.domain_version_readable_min = readable_min.into();
        self
    }

    pub fn version_readable_min(&self) -> Option<BaboonDomainVersion> {
        if self.domain_version_readable_min.is_empty() {
            return self.version_min_compat();
        }
        if self.domain_version_readable_min == self.domain_version {
            None
        } else {
            Some(BaboonDomainVersion::new(&self.domain_identifier, &self.domain_version_readable_min))
        }
    }

    pub fn version_ref(&self) -> BaboonDomainVersion {
        BaboonDomainVersion::new(&self.domain_identifier, &self.domain_version)
    }

    pub fn version_min_compat(&self) -> Option<BaboonDomainVersion> {
        if self.domain_version_min_compat.is_empty() || self.domain_version_min_compat == self.domain_version {
            None
        } else {
            Some(BaboonDomainVersion::new(&self.domain_identifier, &self.domain_version_min_compat))
        }
    }
}

// --- BaboonTypeMeta wire codec ---
//
// Mirrors C#'s static `BaboonTypeMetaCodec` (BaboonTypeMeta.cs:126-214) and Scala's
// `BaboonTypeMetaCodec` (BaboonRuntimeShared.scala:156-214). Used by the facade encode/decode
// entry points to write/read the meta prefix that precedes the payload in both binary and
// JSON envelopes. Lives here (alongside `BaboonTypeMeta`) rather than in `any_opaque.rs`
// because it's part of the facade-level wire surface, not the AnyMeta protocol.

pub mod baboon_type_meta_codec {
    use super::{BaboonTypeMeta, BaboonCodecError};
    use crate::baboon_runtime::bin_tools;
    use std::io::{Read, Write};

    pub const META_VERSION_KEY: &str = "$mv";
    pub const DOMAIN_IDENTIFIER_KEY: &str = "$d";
    pub const DOMAIN_VERSION_KEY: &str = "$v";
    pub const DOMAIN_VERSION_MIN_COMPAT_KEY: &str = "$uv";
    pub const DOMAIN_VERSION_READABLE_KEY: &str = "$rv";
    pub const TYPE_IDENTIFIER_KEY: &str = "$t";

    /// Wire format: `[meta-version:u8][domain:string][version:string][has-min-compat:u8][min-compat?:string][type-id:string]`.
    /// Mirrors C# BaboonTypeMetaCodec.WriteBin / Scala writeBin exactly.
    pub fn write_bin(meta: &BaboonTypeMeta, writer: &mut dyn Write) -> std::io::Result<()> {
        bin_tools::write_byte(writer, BaboonTypeMeta::META_VERSION)?;
        bin_tools::write_string(writer, &meta.domain_identifier)?;
        bin_tools::write_string(writer, &meta.domain_version)?;
        if meta.domain_version == meta.domain_version_min_compat {
            bin_tools::write_byte(writer, 0)?;
        } else {
            bin_tools::write_byte(writer, 1)?;
            bin_tools::write_string(writer, &meta.domain_version_min_compat)?;
        }
        bin_tools::write_string(writer, &meta.type_identifier)
    }

    /// Reads the wire-format prefix. Returns `Ok(None)` when the leading meta-version byte
    /// does not match `META_VERSION_1` (forward-compat: future meta versions). Mirrors C#
    /// `ReadMeta(BinaryReader)` (BaboonTypeMeta.cs:169) returning nullable.
    pub fn read_bin<R: Read>(reader: &mut R) -> Result<Option<BaboonTypeMeta>, BaboonCodecError> {
        let meta_version = bin_tools::read_byte(reader).map_err(|e| {
            BaboonCodecError::decoder_failure_from_box(
                "BaboonTypeMetaCodec.read_bin: failed to read meta-version byte",
                e,
            )
        })?;
        if meta_version != BaboonTypeMeta::META_VERSION_1 {
            return Ok(None);
        }
        let domain_identifier = bin_tools::read_string(reader).map_err(|e| {
            BaboonCodecError::decoder_failure_from_box(
                "BaboonTypeMetaCodec.read_bin: failed to read domain identifier",
                e,
            )
        })?;
        let domain_version = bin_tools::read_string(reader).map_err(|e| {
            BaboonCodecError::decoder_failure_from_box(
                "BaboonTypeMetaCodec.read_bin: failed to read domain version",
                e,
            )
        })?;
        let has_min_compat = bin_tools::read_byte(reader).map_err(|e| {
            BaboonCodecError::decoder_failure_from_box(
                "BaboonTypeMetaCodec.read_bin: failed to read has-min-compat byte",
                e,
            )
        })?;
        // codec-envelope.md §2.1: only 0x00 (elided) and 0x01 (present) are legal; anything else is rejected
        if has_min_compat != 0 && has_min_compat != 1 {
            return Ok(None);
        }
        let domain_version_min_compat = if has_min_compat == 1 {
            bin_tools::read_string(reader).map_err(|e| {
                BaboonCodecError::decoder_failure_from_box(
                    "BaboonTypeMetaCodec.read_bin: failed to read min-compat",
                    e,
                )
            })?
        } else {
            domain_version.clone()
        };
        let type_identifier = bin_tools::read_string(reader).map_err(|e| {
            BaboonCodecError::decoder_failure_from_box(
                "BaboonTypeMetaCodec.read_bin: failed to read type identifier",
                e,
            )
        })?;
        Ok(Some(BaboonTypeMeta::new(
            BaboonTypeMeta::META_VERSION,
            domain_identifier,
            domain_version,
            domain_version_min_compat,
            type_identifier,
        )))
    }

    /// JSON envelope writer. Mirrors C# `BaboonTypeMetaCodec.WriteJson` (BaboonTypeMeta.cs:156)
    /// and Scala `writeMeta(json)` — returns the `{$mv,$d,$v,$t,$uv?}` object without `$c`.
    /// Callers (e.g. `BaboonCodecsFacade::encode_to_json`) are expected to append the content
    /// payload under the `$c` key. Symmetric to `read_meta_json` so the rust runtime exposes
    /// the same shape of public envelope API as the peer backends; closes
    /// `[MFACADE-PR-3-D13]`.
    pub fn write_meta_json(meta: &BaboonTypeMeta) -> serde_json::Map<String, serde_json::Value> {
        let mut envelope = serde_json::Map::new();
        // MFACADE-PR-3: always emit `$mv` as a JSON number so envelopes are
        // self-identifying without out-of-band knowledge (proposal §10.6 (a)).
        envelope.insert(
            META_VERSION_KEY.to_string(),
            serde_json::Value::Number(serde_json::Number::from(BaboonTypeMeta::META_VERSION)),
        );
        envelope.insert(
            DOMAIN_IDENTIFIER_KEY.to_string(),
            serde_json::Value::String(meta.domain_identifier.clone()),
        );
        envelope.insert(
            DOMAIN_VERSION_KEY.to_string(),
            serde_json::Value::String(meta.domain_version.clone()),
        );
        envelope.insert(
            TYPE_IDENTIFIER_KEY.to_string(),
            serde_json::Value::String(meta.type_identifier.clone()),
        );
        if meta.domain_version != meta.domain_version_min_compat
            && !meta.domain_version_min_compat.is_empty()
        {
            envelope.insert(
                DOMAIN_VERSION_MIN_COMPAT_KEY.to_string(),
                serde_json::Value::String(meta.domain_version_min_compat.clone()),
            );
        }
        // `$rv` is elided when it equals the effective `$uv`: unchanged types emit no new bytes
        if !meta.domain_version_readable_min.is_empty()
            && meta.domain_version_readable_min != meta.domain_version_min_compat
        {
            envelope.insert(
                DOMAIN_VERSION_READABLE_KEY.to_string(),
                serde_json::Value::String(meta.domain_version_readable_min.clone()),
            );
        }
        envelope
    }

    /// JSON envelope reader. Mirrors C# `ReadMeta(JToken)` (BaboonTypeMeta.cs:189) and Scala
    /// `readMeta(json)` (BaboonRuntimeShared.scala:197). Returns `Ok(None)` when the input is
    /// not an object, when `$mv` is present but not "1", or when any of `$d`/`$v`/`$t` is missing.
    pub fn read_meta_json(
        json: &serde_json::Value,
    ) -> Result<Option<BaboonTypeMeta>, BaboonCodecError> {
        let obj = match json.as_object() {
            Some(o) => o,
            None => return Ok(None),
        };
        if let Some(mv) = obj.get(META_VERSION_KEY) {
            // MFACADE-PR-3: accept $mv as either a JSON number or a string (back-compat
            // with M28-vintage fixtures); both must equal META_VERSION_1.
            let mv_byte: Option<u8> = match mv {
                serde_json::Value::Number(n) => n.as_u64().and_then(|x| u8::try_from(x).ok()),
                serde_json::Value::String(s) => s.parse::<u8>().ok(),
                _ => None,
            };
            match mv_byte {
                Some(v) if v == BaboonTypeMeta::META_VERSION_1 => {}
                _ => return Ok(None),
            }
        }
        let d = match obj.get(DOMAIN_IDENTIFIER_KEY).and_then(|v| v.as_str()) {
            Some(s) => s,
            None => return Ok(None),
        };
        let v = match obj.get(DOMAIN_VERSION_KEY).and_then(|v| v.as_str()) {
            Some(s) => s,
            None => return Ok(None),
        };
        let t = match obj.get(TYPE_IDENTIFIER_KEY).and_then(|v| v.as_str()) {
            Some(s) => s,
            None => return Ok(None),
        };
        let uv = obj
            .get(DOMAIN_VERSION_MIN_COMPAT_KEY)
            .and_then(|v| v.as_str())
            .unwrap_or(v);
        let rv = obj
            .get(DOMAIN_VERSION_READABLE_KEY)
            .and_then(|v| v.as_str())
            .unwrap_or(uv);
        Ok(Some(
            BaboonTypeMeta::new(BaboonTypeMeta::META_VERSION, d, v, uv, t).with_readable_min(rv),
        ))
    }
}
